lib/db/schema/patch-066-067.sql

```diff
@@ -0,0 +1,81 @@
+SET NAMES utf8mb4 COLLATE utf8mb4_bin;
+
+-- Add an index on sessionTokens::createdAt to make pruning faster.
+CREATE INDEX `sessionTokens_createdAt`
+ON `sessionTokens` (`createdAt`)
+ALGORITHM=INPLACE
+LOCK=NONE;
+
+-- Used to prevent session token pruning from doing a full table scan
+-- as it proceeds further and further through the (very long) backlog
+-- of pruning candidates.
+INSERT INTO dbMetadata (name, value)
+VALUES ('sessionTokensPrunedUntil', 0);
+
+-- Update prune to delete sessionTokens and unverifiedTokens.
+CREATE PROCEDURE `prune_6` (IN `olderThan` BIGINT UNSIGNED)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  SELECT @lockAcquired := GET_LOCK('fxa-auth-server.prune-lock', 3);
+
+  IF @lockAcquired THEN
+    DELETE FROM accountResetTokens WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM passwordForgotTokens WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM passwordChangeTokens WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM unblockCodes WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM signinCodes WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+
+	-- Pruning session tokens is more complicated because we can't
+	-- prune them if there is an associated device record.
+    START TRANSACTION;
+
+    -- Step 1: Find out how far we got on previous iterations.
+    SELECT @pruneFrom := value FROM dbMetadata WHERE name = 'sessionTokensPrunedUntil';
+
+    -- Step 2: Find out how far we will get on this iteration.
+    SELECT @pruneUntil := MAX(createdAt) FROM (
+      SELECT createdAt FROM sessionTokens
+      WHERE createdAt >= @pruneFrom AND createdAt < olderThan
+      AND NOT EXISTS (
+        SELECT sessionTokenId FROM devices
+        WHERE uid = sessionTokens.uid
+        AND sessionTokenId = sessionTokens.tokenId
+      )
+      ORDER BY createdAt
+      LIMIT 10000
+    ) AS prunees;
+
+    -- Step 3: Prune sessionTokens.
+    DELETE FROM sessionTokens
+    WHERE createdAt > @pruneFrom AND createdAt <= @pruneUntil
+    AND NOT EXISTS (
+      SELECT sessionTokenId FROM devices
+      WHERE uid = sessionTokens.uid
+      AND sessionTokenId = sessionTokens.tokenId
+    );
+
+    -- Step 4: Prune unverifiedTokens.
+    DELETE ut
+    FROM unverifiedTokens AS ut
+    LEFT JOIN sessionTokens AS st ON ut.tokenId = st.tokenId
+    LEFT JOIN keyFetchTokens AS kt ON ut.tokenId = kt.tokenId
+    WHERE st.tokenId IS NULL AND kt.tokenId IS NULL;
```

Comment:
This looks like it will still be a full table scan of `unverifiedTokens`.  That table's comparatively small but could still grow indefinitely in theory.

Since our `DELETE FROM sessionTokens` query is no longer using the `ORDER BY` or `LIMIT` clauses that gave us trouble in https://github.com/mozilla/fxa-auth-db-mysql/pull/279#discussion_r138823535, is it now possible to do this `unverifiedTokens` deletion through a join as part of the above query, which would allow it to touch only the rows affected by the to-be-deleted sessionToken ids?

Something like (untested):

```
DELETE st, ut
FROM sessionTokens AS st
LEFT JOIN unverifiedTokens AS ut ON ut.tokenId = st.tokenId
LEFT JOIN devices AS d ON d.uid = st.uid AND d.sessionTokenId = st.tokenId
WHERE st.createdAt > @pruneFrom AND st.createdAt <= @pruneUntil AND d.id IS NULL
```

Reply:
> Since our `DELETE FROM sessionTokens` query is no longer using the `ORDER BY` or `LIMIT` clauses that gave us trouble in https://github.com/mozilla/fxa-auth-db-mysql/pull/279#discussion_r138823535, is it now possible to do this `unverifiedTokens` deletion through a join...

I forgot about that, sorry. Will add (and test the perf).